File: AndroidWebBox/coconut-core/src/main/java/com/sniper/coconut/component/BaseComponent.kt
package com.sniper.coconut.component

import com.sniper.coconut.utils.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * Base Component Class
 *
 * All components should extend this class
 * Provides common functionality and utilities
 */
abstract class BaseComponent : CoconutPlugin {

    protected val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private var _initialized = false

    final override val isInitialized: Boolean
        get() = _initialized

    final override suspend fun init(context: ComponentContext) {
        if (_initialized) {
            Logger.w(name, "Component already initialized")
            return
        }

        Logger.d(name, "Initializing component...")
        onInit(context)
        _initialized = true
        Logger.d(name, "✓ Component initialized")
    }

    /**
     * Called when component is being initialized
     * Override this method to perform component-specific initialization
     *
     * @param context Component context
     */
    protected open suspend fun onInit(context: ComponentContext) {
        // Override in subclasses if needed
    }

    final override suspend fun cleanup() {
        Logger.d(name, "Cleaning up component...")
        onCleanup()
        _initialized = false
        Logger.d(name, "✓ Component cleaned up")
    }

    /**
     * Called when component is being cleaned up
     * Override this method to perform component-specific cleanup
     */
    protected open suspend fun onCleanup() {
        // Override in subclasses if needed
    }

    /**
     * Get parameter value as string
     */
    protected fun getParam(params: JsonObject?, key: String, default: String = ""): String {
        return params?.get(key)?.let {
            when (it) {
                is JsonPrimitive -> if (it.isString) it.content else default
                else -> it.toString()
            }
        } ?: default
    }

    /**
     * Get parameter as int
     */
    protected fun getIntParam(params: JsonObject?, key: String, default: Int = 0): Int {
        return params?.get(key)?.let {
            when (it) {
                is JsonPrimitive -> if (it is JsonPrimitive) it.content.toIntOrNull() ?: default else default
                else -> default
            }
        } ?: default
    }

    /**
     * Get parameter as boolean
     */
    protected fun getBoolParam(params: JsonObject?, key: String, default: Boolean = false): Boolean {
        return params?.get(key)?.let {
            when (it) {
                is JsonPrimitive -> if (it is JsonPrimitive) it.content.toBooleanStrictOrNull() ?: default else default
                else -> default
            }
        } ?: default
    }

    /**
     * Build success result
     *
     * @param data Result data
     * @return JsonElement with success format
     */
    protected fun success(data: JsonElement? = null): JsonElement {
        return buildJsonObject {
            put("code", JsonPrimitive("000000"))
            put("message", JsonPrimitive("success"))
            data?.let { put("data", it) }
        }
    }

    /**
     * Build error result
     *
     * @param code Error code (string)
     * @param message Error message
     * @return JsonElement with error format
     */
    protected fun error(code: String, message: String): JsonElement {
        return buildJsonObject {
            put("code", JsonPrimitive(code))
            put("message", JsonPrimitive(message))
        }
    }

    /**
     * Build function not supported error
     */
    protected fun functionNotSupportedError(function: String): JsonElement {
        return error("900002", "Function not supported: $function")
    }
}

File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Core/BaseComponent.swift
import Foundation
import UIKit

open class BaseComponent: CoconutPlugin {

    public init() {}

    open var name: String { "" }
    open var version: String { "1.0.0" }
    open var pluginDescription: String { "" }
    open var dependencies: [String] { [] }

    private var _initialized = false
    public var isInitialized: Bool { _initialized }

    open func onInit(context: ComponentContext) async {}
    open func handle(function: String, params: [String: Any]?) async throws -> [String: Any] {
        try functionNotSupportedError(function)
    }
    open func onCleanup() async {}

    public final func initComponent(context: ComponentContext) async throws {
        guard !_initialized else {
            Logger.shared.w(name, "Component already initialized")
            return
        }
        Logger.shared.d(name, "Initializing component...")
        await onInit(context: context)
        _initialized = true
        Logger.shared.i(name, "✓ Component initialized")
    }

    public final func cleanup() async {
        Logger.shared.d(name, "Cleaning up component...")
        await onCleanup()
        _initialized = false
        Logger.shared.d(name, "✓ Component cleaned up")
    }

    // MARK: - Helper Methods

    public func getParam(_ params: [String: Any]?, _ key: String, _ defaultValue: String = "") -> String {
        guard let value = params?[key] else { return defaultValue }
        if let str = value as? String { return str }
        return "\(value)"
    }

    public func getIntParam(_ params: [String: Any]?, _ key: String, _ defaultValue: Int = 0) -> Int {
        guard let value = params?[key] else { return defaultValue }
        if let int = value as? Int { return int }
        if let str = value as? String, let int = Int(str) { return int }
        if let double = value as? Double { return Int(double) }
        return defaultValue
    }

    public func getLongParam(_ params: [String: Any]?, _ key: String, _ defaultValue: Int64 = 0) -> Int64 {
        guard let value = params?[key] else { return defaultValue }
        if let int = value as? Int64 { return int }
        if let int = value as? Int { return Int64(int) }
        if let str = value as? String, let int = Int64(str) { return int }
        if let double = value as? Double { return Int64(double) }
        return defaultValue
    }

    public func getBoolParam(_ params: [String: Any]?, _ key: String, _ defaultValue: Bool = false) -> Bool {
        guard let value = params?[key] else { return defaultValue }
        if let bool = value as? Bool { return bool }
        if let str = value as? String { return str.lowercased() == "true" }
        return defaultValue
    }

    /// Build success result - returns data dict directly.
    /// Bridge layer handles code/message at top level.
    public func success(_ data: [String: Any]? = nil) -> [String: Any] {
        return data ?? [:]
    }

    /// Throw component error to be caught by Bridge layer
    public func error(_ code: String, _ message: String) throws -> Never {
        throw ComponentException(code: code, message: message)
    }

    /// Throw function not supported error
    public func functionNotSupportedError(_ function: String) throws -> Never {
        throw ComponentException(code: ErrorCode.UNKNOWN_FUNCTION, message: "Function not supported: \(function)")
    }
}
